### ISSUE TITLE
Files opened within a container will not appear in subsequent containers if its path changes

### ISSUE THREAD

**Issue by funkyhippo** [docker/for-win#6200]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 4E942A84-2689-4A8E-9582-3617ED83B322/20200401215927

### Expected behavior

When a shared file is opened within a container, changes in its host path (ie. one of its parent directories gets renamed) should not affect any subsequent containers trying to open that file.

### Actual behavior

If its path changes, the shared file will no longer appear in any subsequent containers.

### Information

  - Windows Version: 1909 (OS Build 18363.720)
  - Docker Desktop Version: 2.2.0.3 (but reproducible in 2.2.0.4 and 2.2.0.5)
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No

The problem appears to be from 2.2.0.3 onward (I've tested 2.2.0.3, 2.2.0.4, and 2.2.0.5 from https://github.com/docker/for-win/issues/5530). 2.1.0.5 does not exhibit this issue. I could not test 2.2.0.0 since its download was pulled, but I think it's fair to assume the problem's related to the new file sharing changes.

### Steps to reproduce the behavior

  1. Create a directory in any arbitrary location (eg. `C:/test`)
  2. Create a file within (eg. `file`)
  3. Open the file within a container (eg. `docker run --rm -v C:/test/:/test alpine /bin/cat /test/file`)
  4. Rename the directory (eg. `C:/test` -> `C:/renamed`)
  5. Open the file within a container again (eg. `docker run --rm -v C:/renamed/:/renamed alpine /bin/cat /renamed/file`)

The last step will fail:
```
C:\renamed>docker run --rm -v C:/renamed/:/renamed alpine /bin/cat /renamed/file
cat: can't open '/renamed/file': No such file or directory
```
`ls` in the directory will also give:
```
C:\renamed>docker run --rm -v C:/renamed/:/renamed alpine /bin/ls /renamed
ls: /renamed/file: No such file or directory
```

_This is a repost of this issue here: https://github.com/docker/compose/issues/7327; I had originally thought it was a compose issue but upon further investigation, it seems to be a Docker Desktop issue._

**Comment by stephen-turner** [comment]
Interesting. We'll test this case.

**Comment by stephen-turner** [comment]
Confirmed, I have reproduced this. Strangely, for me step 5 worked once but when I did it a second time it failed.

@djs55 Could you take a look please? If you can't reproduce it, and you want to do a Zoom with me or get me to produce any traces, I can do that.

**Comment by djs55** [comment]
@funkyhippo Thanks for your report and clear repro instructions. I was able to reproduce this and track it down to a stale cache entry in the VM: when we stop watching a filesystem path `/a/b/c` we flush the caches for all the children `/a/b/c/...` but we didn't flush the entries for parents `/a`, `/a/b`. If you have the time to experiment, I've made a testing build with a proposed fix: https://download-stage.docker.com/win/edge/44128/Docker%20Desktop%20Installer.exe If you have a chance to try it, let me know if it works any better or not. Thanks again for the clear report!

**Comment by funkyhippo** [comment]
Hey @djs55, thanks for the quick turnaround! I played around with the test build, and it seems to be working a bit better. I can't reproduce it anymore with the steps I outlined, which is great. Unfortunately, I'm still noticing occasional instances where it fails but I can't reliably reproduce it so this isn't particularly helpful for you.

At the moment, the closest I've gotten to being able to reliably reproduce it is by using `docker-compose`, by following the steps in this repo I originally made for this issue: https://github.com/funkyhippo/docker-min. This isn't a great update since there's a chance it could be a compose issue here, but I figured I'd leave this here just in case. 

On a slightly unrelated note, one thing that I did notice with this build was that I'd consistently get this error when I navigated to the mounted directory within a container: 
```shell
sh: getcwd: No such file or directory
```
Here's my full console output:
```shell
C:\>mkdir test

C:\>docker run -it --rm -v C:/test:/test alpine sh
/ # cd test
sh: getcwd: No such file or directory
(unknown) # ls
sh: getcwd: No such file or directory
(unknown) # cd ..
/ # ls
bin    etc    lib    mnt    proc   run    srv    test   usr
dev    home   media  opt    root   sbin   sys    tmp    var
/ #
```
I'm not sure if this build included any of the other edge updates, but I figured it'd be worth mentioning.

**Comment by djs55** [comment]
@funkyhippo thanks for the feedback -- it looks like I invalidated the wrong cache entries, not fully fixing the issue and causing the `getcwd: No such file or directory` bug in the process. I've made a more targeted fix which hopefully works better. If you'd like to try it it's here: https://download-stage.docker.com/win/edge/44226/Docker%20Desktop%20Installer.exe . With this build I was able to use your compose example without a failure:
```
PS C:\Users\djs\bugs\6200\docker-min> docker-compose up
Creating network "docker-min_default" with the default driver
Building test
Step 1/1 : FROM python:3-alpine
3-alpine: Pulling from library/python
aad63a933944: Pull complete
f229563217f5: Pull complete
d999dd4b9386: Pull complete
d40444ccd481: Pull complete
d7d60c647873: Pull complete
Digest: sha256:9bb7c74b43a0eba7a7cbda466dafac5e082ad6549d343740ed253788547701e2
Successfully tagged docker-min_test:latest
WARNING: Image for service test was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.Creating docker-min_test_1 ... done
Attaching to docker-min_test_1
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
Gracefully stopping... (press Ctrl+C again to force)
Stopping docker-min_test_1 ...
Killing docker-min_test_1  ... done
PS C:\Users\djs\bugs\6200\docker-min> cd ..
PS C:\Users\djs\bugs\6200> mv .\docker-min .\docker-min2
PS C:\Users\djs\bugs\6200> cd .\docker-min2\
PS C:\Users\djs\bugs\6200\docker-min2> docker-compose up
Creating network "docker-min2_default" with the default driver
Building test
Successfully tagged docker-min2_test:latest
WARNING: Image for service test was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.Creating docker-min2_test_1 ... done
Attaching to docker-min2_test_1
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
Gracefully stopping... (press Ctrl+C again to force)
Stopping docker-min2_test_1 ...
Killing docker-min2_test_1  ... done
PS C:\Users\djs\bugs\6200\docker-min2> cd ..
PS C:\Users\djs\bugs\6200> mv .\docker-min2 .\docker-min3
PS C:\Users\djs\bugs\6200> cd .\docker-min3\
PS C:\Users\djs\bugs\6200\docker-min3> docker-compose up

Successfully built d5e5ad4a4fc0
Successfully tagged docker-min3_test:latest
WARNING: Image for service test was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.Creating docker-min3_test_1 ... done
Attaching to docker-min3_test_1
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
Gracefully stopping... (press Ctrl+C again to force)
PS C:\Users\djs\bugs\6200\docker-min3> cd ..
PS C:\Users\djs\bugs\6200> mv .\docker-min3 .\docker-min4
PS C:\Users\djs\bugs\6200> cd .\docker-min4\
PS C:\Users\djs\bugs\6200\docker-min4> docker-compose up
Creating network "docker-min4_default" with the default driver
Building test
Step 1/1 : FROM python:3-alpine
 ---> d5e5ad4a4fc0

Successfully built d5e5ad4a4fc0
WARNING: Image for service test was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.Creating docker-min4_test_1 ... done
Attaching to docker-min4_test_1
test_1  | Sleeping..
test_1  | Sleeping..
test_1  | Sleeping..
Gracefully stopping... (press Ctrl+C again to force)
Stopping docker-min4_test_1 ...
Killing docker-min4_test_1  ... done
PS C:\Users\djs\bugs\6200\docker-min4> cd ..
PS C:\Users\djs\bugs\6200> cd .\docker-min5\
PS C:\Users\djs\bugs\6200\docker-min5> docker-compose up
Creating network "docker-min5_default" with the default driver
Building test
Step 1/1 : FROM python:3-alpine
 ---> d5e5ad4a4fc0
Successfully built d5e5ad4a4fc0
test_1  | Sleeping..
Stopping docker-min5_test_1 ...
```
On previous builds this failed quite reliably for me. Thanks again for all your help so far with this issue!

**Comment by funkyhippo** [comment]
@djs55 beauty! That build fixes all the issues I'm seeing on my end. 

I'm glad I was able to help out, and thanks for all your work :)

**Comment by stephen-turner** [comment]
Fixed in 2.3, closing this ticket.
